include/sparrow/buffer.hpp

```diff
@@ -686,6 +702,132 @@ namespace sparrow
         erase_at_end(get_data().p_begin);
     }
 
+    template <class T>
+    constexpr auto buffer<T>::insert(const_iterator pos, const T& value) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        return emplace(pos, value);
+    }
+
+    template <class T>
+    constexpr auto buffer<T>::insert(const_iterator pos, T&& value) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        return emplace(pos, std::move(value));
+    }
+
+    template <class T>
+    constexpr auto buffer<T>::insert(const_iterator pos, size_type count, const T& value) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        const difference_type __offset = pos - cbegin();
+        pointer ptr = get_data().p_begin + __offset;
+        if (count != 0)
+        {
+            const size_type sz = size();
+            const size_type off = ptr - get_data().p_begin;
+            if (const size_type new_size = sz + count; new_size > capacity())
+            {
+                reserve(new_size);
+            }
+            auto& data = get_data();
+            const auto new_ptr = data.p_begin + off;
+            std::move_backward(new_ptr, data.p_end, data.p_end + count);
+            std::fill_n(new_ptr, count, value);
+            data.p_end += count;
+        }
+        return begin() + __offset;
+    }
+
+    template <class T>
+    template <class InputIt>
+    constexpr auto buffer<T>::insert(const_iterator pos, InputIt first, InputIt last) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        const size_type num_elements = last - first;
+        const size_type new_size = size() + num_elements;
+        const size_type offset = pos - cbegin();
+        reserve(new_size);
+        auto& data = get_data();
+        const pointer p = data.p_begin + offset;
+        std::move_backward(p, data.p_end, data.p_end + num_elements);
+        std::copy(first, last, p);
+        data.p_end += num_elements;
+        return iterator(p);
+    }
+
+    template <class T>
+    constexpr auto buffer<T>::insert(const_iterator pos, std::initializer_list<T> ilist) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        return insert(pos, ilist.begin(), ilist.end());
+    }
+
+    template <class T>
+    template <class... Args>
+    constexpr buffer<T>::iterator buffer<T>::emplace(const_iterator pos, Args&&... args)
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        const size_type sz = size();
+        const size_type off = pos - cbegin();
+        if (sz <= capacity())
+        {
+            reserve(sz + 1);
+        }
+        pointer p = get_data().p_begin + off;
+        if (p != get_data().p_end)
+        {
+            alloc_traits::construct(get_allocator(), get_data().p_end, std::move(*(get_data().p_end - 1)));
```

Comment:
I'm confused, why are both construct and move_backward on the same element necessary here?

Reply:
The two operations are doing different things:

1. `alloc_traits::construct(get_allocator(), get_data().p_end, std::move(*(get_data().p_end - 1)));` is constructing a new object at the location `get_data().p_end` by moving the object that is currently at `get_data().p_end - 1`. After this operation, the object at `get_data().p_end - 1` is in a valid but unspecified state.

2. `std::move_backward(p, get_data().p_end - 1, get_data().p_end);` is moving the objects in the range `[p, get_data().p_end - 1)` to the range `[p + 1, get_data().p_end)`. This effectively shifts the objects one position to the right.

The reason both are necessary is because the first operation creates space for the second operation to move the objects into. Without the first operation, the second operation would overwrite the last object in the range.